docs/language/informal/toplevel-inference.md

```diff
@@ -0,0 +1,311 @@
+﻿# Toplevel inference
+
+Owner: leafp@google.com
+
+Status: Proposal
+
+
+## Top level inference overview
+
+Top level inference has two phases:
+
+1. **Method override inference**
+    * If you omit a return type or parameter type from an overridden method,
+    inference will try to fill in the missing type using the signature of the
+    method you are overriding.
+2. **Static variable and field inference** 
+    * If you omit the type of a field, setter, or getter, which overrides a
+   corresponding member of a superclass, then inference will try to fill in the
+   missing type using the type of the corresponding member of the superclass.
+    * Otherwise, declarations of static variables and fields that omit a type
+   will be inferred from their initializer if present.
```

Comment:
> Hence, I wanted to make that commitment visible---if my new setter in this new class is to get a type from somewhere out there, it ought to be stated as a type annotation. In contrast, relying on an initializer expression in an overridden instance variable declaration could mean that several libraries suddenly get different signatures on several methods, suddenly giving lots of call sites different typing properties,

The same is true of changing the annotated type on a field, for what it's worth.  I hear what you're saying.  I'm not super keen on override inference in general for this reason.

However, we did enunciate a principle when we discussed this in August that there should be no difference in the treatment of inferred types vs annotated types.  Splitting out override inference from initializer inference from fields seems to violate that, and seems like it would lead to odd errors.  For example, this code:

```
class A {
  var x = 3;
}

class B extends A {
  var x;
}
```

Would generate an override error for `x`, since `B.x` would get `dynamic` from override inference before `A.x` got inferred to `int`.  Even worse:

```
class A {
  final var x = 3;
}

class B extends A {
  final var x;
}
```
would generate no error, but would silently infer `B.x` as `dynamic`, while `A.x` would get inferred as `int`.  It feels weird to me.